File: day14/typescript/day14.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

import { part2 as knot } from '../../day10/typescript/day10';
import { part2 as connected, Graph } from '../../day12/typescript/day12';

type Coord = R.KeyValuePair<number, number>;
type Grid = number[];

const binary = (hex: string) => {
    const i = parseInt(hex, 16);
    const bin = i.toString(2).padStart(4, '0');
    return R.map(parseInt, Array.from(bin));
};

const makeGrid = (input: string) => {
    const hashes = R.map(knot, R.times(n => `${input}-${n}`, 128));
    return R.chain(R.compose(R.chain(binary), Array.from), hashes);
};

const part1 = R.sum;

const neighbors = (coord: string) => {
    const [x, y] = JSON.parse(coord);
    return R.map(JSON.stringify, [
        [x + 1, y],
        [x, y + 1],
        [x - 1, y],
        [x, y - 1],
    ]);
};

const occupied = R.compose<Grid, number[][], number[][], string[]>(
    R.map<number[], string>(([_, x, y]) => JSON.stringify([x, y])),
    R.filter<number[]>(([v, x, y]) => v === 1),
    R.addIndex<number, number[]>(R.map)(
        (val, idx) => [val, Math.floor(idx / 128), idx % 128],
    ),
);

const makeGraph = (cs: Set<string>) => R.reduce<string, Graph>(
    (acc, curr) => R.set(
        R.lensProp(curr),
        R.filter(c => cs.has(c), neighbors(curr)),
        acc,
    ),
    {},
    Array.from(cs),
);

const part2 = R.compose(
    connected,
    makeGraph,
    o => new Set(o),
    occupied,
);

(async () => {
    const input = await promisify(readFile)('day14/input.txt', 'utf8');
    const grid = makeGrid(input);

    console.log(part1(grid));
    console.log(part2(grid));
})();

File: day14/haskell/day14.hs
import qualified Day10.Haskell.Day10 as D10
import qualified Day12.Haskell.Day12 as D12

import Numeric (showIntAtBase, showInt)
import Control.Arrow (first)
import Data.Char (intToDigit, digitToInt)
import qualified Data.Map as M
import qualified Data.Set as S

type Grid = [Int]
type Coord = (Int, Int)
type Graph = M.Map Coord [Coord]

binary :: Char -> [Int]
binary s = map digitToInt . D10.lpad 4 '0' $ showIntAtBase 2 intToDigit n []
    where n = digitToInt s

makeGrid :: String -> Grid
makeGrid s = concatMap f [0..127]
    where f = concatMap binary . D10.part2 . ((s ++ "-") ++) . flip showInt []

part1 :: Grid -> Int
part1 = sum

neighbors :: Coord -> [Coord]
neighbors (x, y) = [(x + 1, y), (x, y + 1), (x - 1, y), (x, y - 1)] 

occupied :: Grid -> [Coord]
occupied = map fst . filter ((== 1) . snd) . map (first (`divMod` 128)) . zip [0..]

makeGraph :: S.Set Coord -> Graph
makeGraph cs = foldl f M.empty cs
    where f a c = M.insert c (filter (`S.member` cs) (neighbors c)) a

part2 :: Grid -> Int
part2 = D12.part2 . makeGraph . S.fromList . occupied

main :: IO ()
main = do
    grid <- makeGrid <$> readFile "day14/input.txt"
    print . part1 $ grid
    print . part2 $ grid